edgedb-protocol/src/query_arg.rs

```diff
@@ -526,3 +526,109 @@ implement_tuple! {9, T0, T1, T2, T3, T4, T5, T6, T7, T8, }
 implement_tuple! {10, T0, T1, T2, T3, T4, T5, T6, T7, T8, T9, }
 implement_tuple! {11, T0, T1, T2, T3, T4, T5, T6, T7, T8, T9, T10, }
 implement_tuple! {12, T0, T1, T2, T3, T4, T5, T6, T7, T8, T9, T10, T11, }
+
+// This supertype allows user to provide both
+// Into<Value>, Option<Into<Value>>, Vec<Into<Value>>, Option<Vec<Into<Value>>>
+// types in HashMap
+pub struct UserValue(Option<Value>);
+
+impl<V: Into<Value>> From<V> for UserValue {
+	fn from(value: V) -> Self {
+		UserValue(Some(value.into()))
+	}
+}
+impl<V: Into<Value>> From<Option<V>> for UserValue
+where
+	Value: From<V>
+{
+	fn from(value: Option<V>) -> Self {
+		UserValue(value.map(Value::from))
+	}
+}
+impl<V: Into<Value>> From<Vec<V>> for UserValue
+where
+	Value: From<V>
+{
+	fn from(value: Vec<V>) -> Self {
+		UserValue(Some(Value::Array(value.into_iter().map(Value::from).collect())))
+	}
+}
+impl<V: Into<Value>> From<Option<Vec<V>>> for UserValue
+where
+	Value: From<V>
+{
+	fn from(value: Option<Vec<V>>) -> Self {
+		let mapped = value.map(|value| Value::Array(value.into_iter().map(Value::from).collect()));
+		UserValue(mapped)
+	}
+}
+impl From<UserValue> for Option<Value> {
+	fn from(value: UserValue) -> Self {
+		value.0
+	}
+}
+
+use std::collections::HashMap;
+impl QueryArgs for HashMap<&str, UserValue> {
+	fn encode(&self, encoder: &mut Encoder) -> Result<(), Error> {
+		let target_shape = {
```

Comment:
This need special case for zero provided arguments in hashmap.

```rs
if self.len() == 0 {
    return ().encode(encoder);
}
```

Would this be enough?